src/contrib/php_array_api.h

```diff
@@ -420,7 +420,17 @@ char *php_array_zval_to_string(zval *z, int *plen, zend_bool *pfree) {
  * zval *php_array_fetchz_array(zval *zarr, zval *key)
  */
 static inline zval *php_array_zval_to_array(zval *zarr) {
-	return (zarr && (Z_TYPE_P(zarr) == IS_ARRAY)) ? zarr : NULL;
+try_again:
+	if (!zarr) { return NULL; }
+	switch (Z_TYPE_P(zarr)) {
+		case IS_REFERENCE:
+			ZVAL_DEREF(zarr);
+			goto try_again;
```

Comment:
I realize this is the same logic from https://github.com/mongodb/mongo-php-driver/commit/c68cdcda7654a630b6da52d99f7ce5e2b3fdabfc, but I'm curious this could be replaced with a fall-through.

[Dereferencing and Unwrapping](https://www.phpinternalsbook.com/php7/zvals/references.html#dereferencing-and-unwrapping) includes the following code example:

```c
zval *val;
ZEND_HASH_FOREACH_VAL(ht, val) {
    ZVAL_DEREF(val);

    /* Do something with val, now a guaranteed non-reference. */
} ZEND_HASH_FOREACH_END();
```

That implies that dereferencing would never yield another `IS_REFERENCE` type. That is also supported by this earlier text in [Initializing references](https://www.phpinternalsbook.com/php7/zvals/references.html#initializing-references):

> ```c
> zval *zv = /* ... */;
> ZVAL_MAKE_REF(zv);
> ```
> 
> If zv was already a reference, this does nothing. It if wasn’t a reference yet, this will change zv into a reference and set its initial value to the old value of zv.

It's a bit too early for me to dive through the macros in [zend_types.h](https://heap.space/xref/PHP-8.1/Zend/zend_types.h) to confirm this.

Reply:
Looking through [php-src](https://github.com/search?q=repo%3Aphp%2Fphp-src%20ZVAL_DEREF&type=code), I don't see any examples of `ZVAL_DEREF()` being called iteratively. It tends to precede zval access, and code relies on it being a no-op for non-references. In that case, perhaps you could unconditionally call `ZVAL_DEREF(zarr)` after the initial `NULL` check and eliminate the `goto`.

Likewise for the original patch in https://github.com/mongodb/mongo-php-driver/commit/c68cdcda7654a630b6da52d99f7ce5e2b3fdabfc.

If you'd rather not make this change now, I'd consider it for the upstream contribution to [sgolemon/php-array-api](https://github.com/sgolemon/php-array-api/).

Reply:
Looking at this, I don't think that will work. After calling `ZVAL_DEREF` we still have to check the type of the dereferenced zval before accessing the ce, which wouldn't work with a fallthrough. As much as I hate the `goto` way of doing things, the only other option I see is a `while (Z_TYPE_P(z) == IS_REFERENCE)`, which is equally bad.

Reply:
Removed the `goto` after all by calling `ZVAL_DEREF` unconditionally after a `NULL` check.